ZOOKEEPER-3630: Autodetection of openssl during ZooKeeper C client build

symat opened the pull request:
**Thanks for @ztzg for raising the issue and suggesting the solution!**

In this patch we enhance the way how the openssl library is found during C client build. I introduced  and documented a new build parameter for `configure` (`--with-openssl=...`), `cmake` (`-D WITH_OPENSSL=...`) and `mvn` (`-Dc-client-openssl=...`), so independent of the build tool, the build will work the same way:
- By default, the OpenSSL library will be autodetected. If the library is found, then the C-client will be compiled with SSL support, otherwise we get a warning message, but the build will continue without SSL support. The SSL related unit tests will be skipped as well.
- you can explicitly disable the SSL support (e.g. `-Dc-client-openssl=no`)
- or you can specify an alternative path to look for the openssl library (e.g. `-Dc-client-openssl=/path/to/openssl`)

I tested the patch
- using mvn on linux
- using make on linux
- using cmake on linux and on windows

In addition, I also added the openssl dependencies to the dev docker image and copied the OpenSSL license to the C client LICENSE file.

symat commented:
> (I am not a C expert)

(me neither :) )

> are we compiling statically openssl lib ? 

I think we basically link together all the dependent libs, when building the command line client. At least, I was able to connect to a ZK server using the C command line client with SSL, even when I executed it in a docker container with no openssl installed. So the openssl is only compile-time dependency, but it will end up in the ZooKeeper client binary.

We are also building the ZooKeeper libraries (they are used e.g. by the python binding). Based on the README file in the c-client folder, we can compile both static and shared libraries with autotools, but with cmake we only build static and the shared library function is not implemented yet. Unfortunately I am also quite inexperienced in the C world, so I can not really verify these statements :)

eolivelli commented:
If we are bundling the c client in the binary release package we must handle LICENSE files properly

symat commented:
that is true, good catch, I totally missed that
(it will land in the `apache-zookeeper-<version>-lib.tar.gz`)
How should we handle that? We already have a LICENSE.txt in that tar.gz file, containing a normal Apache license.

eolivelli commented:
I am testing with the built in docker environment (dev/docker/run.sh) but openssl is not available

```

checking openssl/ssl.h usability... no
checking openssl/ssl.h presence... no
checking for openssl/ssl.h... no
configure: WARNING: cannot build SSL support -- openssl not found
```

Do you mind updating the Dockerfile with the download o a known openssl version ?
As we are linking statically we can choose the version we prefer.

symat commented:
> Do you mind updating the Dockerfile with the download o a known openssl version ? 
sure, I will do that and push here soon

> As we are linking statically we can choose the version we prefer
I just checked, but unfortunately there is still no stable release with the new OpenSsl which is using Apache license... so we will have to stick to the old ones and add the license. I will push the OpenSSL license as well and incorporate it to the `apache-zookeeper-<version>-lib.tar.gz` file. Is is OK if I add a new file (e.g. `LICENSE_openssl.txt` ,  or should I put to the end of the current `LICENSE.txt` file?)

eolivelli commented:
The LICENSE file should only go into the binary tarbal of the C client, because here it is the only place where we are redistributing it

symat commented:
by binary tarball you mean `apache-zookeeper-<version>-bin.tar.gz` ? that file contains only java (+docs) now.
I had a PR where I originally wanted to put the C CLI binary and compiled C client libraries into the `apache-zookeeper-<version>-bin.tar.gz` file, but the conclusion was to create a separate file for it, called `apache-zookeeper-<version>-lib.tar.gz`, following the way how it was distributed in 3.4.

symat commented:
this is what we have now in the `apache-zookeeper-3.6.0-SNAPSHOT-lib.tar.gz` artifact:
```
root@9184f8d4a7ca:/git/zookeeper/zookeeper-assembly/target# tar xzvf ./apache-zookeeper-3.6.0-SNAPSHOT-lib.tar.gz
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/zookeeper.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/proto.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/zookeeper_version.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/zookeeper.jute.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/zookeeper_log.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/include/zookeeper/recordio.h
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_st.a
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_mt.la
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_mt.so.2
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_st.so.2
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_mt.so
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_st.so
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_mt.so.2.0.0
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_mt.a
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_st.la
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/lib/libzookeeper_st.so.2.0.0
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/bin/cli_mt
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/bin/cli_st
apache-zookeeper-3.6.0-SNAPSHOT-lib/usr/bin/load_gen
apache-zookeeper-3.6.0-SNAPSHOT-lib/LICENSE.txt
```

eolivelli commented:
Ok I thought we had done the separate package.

Let's put openssl license in that file

symat commented:
In the last commit I added the necessary packages to the Dockerfile, now the build in docker shows:
```
checking openssl/ssl.h usability... yes
checking openssl/ssl.h presence... yes
checking for openssl/ssl.h... yes
checking for SSL_CTX_new in -lssl... yes
configure: building with SSL support
```

I also extended the C client LICENSE file with the OpenSSL license, and changed the assembly config that the `...-lib.tar.gz` file contains this license file and not only the generic Apache license we have for the java code.

symat commented:
@eolivelli what do you think about the changes?

(I think it would worth to merge this PR as it would stabilize the jenkins builds and also make the build easier for the users - so it also might be good to put to 3.6.0, so it would be easier to test it)

@nkalmar - could you please also take a look? (as it is also maven related)

nkalmar commented:
I agree on pushing this to 3.6.0. 
@eolivelli PTAL

ztzg commented:
Very happy to see this in `master`!

FYI, I have now rebased and "fully aligned" my SASL PR on top of it: https://github.com/apache/zookeeper/pull/1134#issuecomment-565808762
